Spanner/src/SpannerClient.php

```diff
@@ -325,6 +376,44 @@ public function instance($name, array $instance = [])
             $this->returnInt64AsObject,
             $instance
         );
+
+        if (!isset($this->config['apiEndpoint']) &&
+            getenv('GOOGLE_CLOUD_ENABLE_RESOURCE_BASED_ROUTING')
+        ) {
+            if (!isset($this->endpoints[$instanceObject->name()])) {
+                $instanceInfo = $instanceObject->info([
+                    'fieldMask' => new FieldMask(['paths' => ['endpoint_uris']])
+                ]);
+                $this->endpoints[$instanceObject->name()] = $instanceInfo['endpointUris'];
+            }
+
+            if (!is_null($apiEndpoint = $this->endpoint($instanceObject->name()))) {
+                $connection = $this->createConnection(['apiEndpoint' => $apiEndpoint]);
+                list($lroConnection, $lroCallables) = $this->createLroProperties($connection);
+                $instanceObject->connectionUpdate($connection, $lroConnection, $lroCallables);
+            }
+        }
+
+        return $instanceObject;
+    }
+
+    /**
+     * Return endpoint for the specified instance, or null if the list for this
+     * instance is empty.
+     *
+     * Example:
+     * ```
+     * if (!is_null($apiEndpoint = $this->endpoint($instance->name()))) {
+     *     // do someting with $apiEndpoint
+     * }
+     * ```
+     *
+     * @return string|null
+     */
+    private function endpoint($instanceName)
+    {
+        return isset($this->endpoints[$instanceName]) && count($this->endpoints[$instanceName])
+            ? $this->endpoints[$instanceName][0] : null;
```

Comment:
I'd move all of the endpoint reading logic into this function

Reply:
The method is designed to return the endpoint or signal its absence. Design
`!isset($this->endpoints[$instanceObject->name()])`
in [on SpannerClient.php line 383](https://github.com/q-logic/google-cloud-php/pull/4/files#diff-fbcc06d551e59c43ae44dc8f92b3f84fR383) does not check the size of the list but only the fact that for this instance-name endpoint is already was requested by the presence of the key in the array. I believe that SpannerClient::endpoint performs similar but other tasks and it would not be right to interfere with them in one method.

**Example:**
The endpoint list is requested for ".../instance-name", an empty list is returned - it is stored in the `$this->endpoints` field. If instead of the specified design
`!isset($this->endpoints[$instanceObject->name()])`
use
`is_null($this->endpoint(".../instance-name"))`  - since the method returns null, a `getInstance` call will be made again. This is an undesirable behavior.

For this reason, I still would not change the current decision. Or maybe I didn’t understand this remark correctly.